sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,645 @@
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Above we see an example of how to register a PSK and an ECC credential.
+ *
+ * First, we need to include the header file for the API.
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * #include "net/credman.h"
+ *
+ * int main(void)
+ * {
+ *     credman_credential_t psk_credential = {
+ *         .type = CREDMAN_TYPE_PSK,
+ *         .tag = SOCK_DTLS_SERVER_TAG,
+ *         .params = {
+ *             .psk = {
+ *                 .key = { .s = psk_key, .len = sizeof(psk_key), },
+ *                 .id = { .s = psk_id, .len = sizeof(psk_id), },
+ *             },
+ *         },
+ *     };
+ *
+ *     [...]
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * We tell [credman](@ref net_credman) which credential to add by filling in
+ * the credentials information in a struct @ref credman_credential_t. For
+ * PSK credentials, we use enum @ref CREDMAN_TYPE_PSK for the
+ * [type](@ref credman_credential_t::type).
+ *
+ * Next, we must assign a [tag](@ref credman_tag_t) for the credential. Tags
+ * are unsigned integer value that is used to identify which DTLS sock has
+ * access to which credential. Each DTLS sock will also be assigned a tag.
+ * As a result, a sock can only use credentials that have the same tag as
+ * its assigned tag.
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * if (credman_add(&psk_credential) < 0) {
+ *     puts("Error cannot add credential");
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * After credential information is filled, we can add it to the credential
+ * pool using @ref credman_add().
+ *
+ * For adding credentials of other types you can follow the steps above except
+ * credman_credential_t::type and credman_credential_t::params depend on the
+ * type of credential used.
+ *
+ * ### Server Operation
+ *
+ * After credentials are added, we can start the server.
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * #include <stdio.h>
+ *
+ * #include "net/sock/dtls.h"
+ *
+ * #define SOCK_DTLS_SERVER_TAG (10)
+ *
+ * int main(void)
+ * {
+ *     // Add credentials
+ *     // [...]
+ *
+ *     // initialize server
+ *     sock_udp_t udp_sock;
+ *     sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+ *     local.port = 20220;
+ *     if (sock_udp_create(&udp_sock, &local, NULL, 0) < 0) {
+ *         puts("Error creating UDP sock");
+ *         return -1;
+ *     }
+ *
+ *     sock_dtls_t dtls_sock;
+ *     if (sock_dtls_create(&dtls_sock, &udp_sock,
+ *                          SOCK_DTLS_SERVER_TAG,
+ *                          SOCK_DTLS_1_2, SOCK_DTLS_SERVER) < 0) {
+ *         puts("Error creating DTLS sock");
+ *         return -1;
+ *     }
+ *
+ *     while (1) {
+ *         int res;
+ *         char buf[128];
+ *         sock_dtls_session_t session;
+ *
+ *         res = sock_dtls_recv(&dtls_sock, &session, buf, sizeof(buf),
+ *                              SOCK_NO_TIMEOUT);
+ *         if (res > 0) {
+ *             printf("Received %d bytes: %*.s\n", res, res, buf);
```

Comment:
This is not a good idea if `buf` is not a string, since then there can be a lot of nasty characters in `buf`.

Reply:
Only if you assume that the client you provide in this example is the one this server is connected with. I don't think its good advertisement to put such a security risk into the server example ;-). People will copy this code and expect it to work.

Reply:
Changed to only print byte length received/sent instead of the content of `buf`.